# Data Exploration and Visual Exploration Lab

In this lab, you will practice an exploratory workflow using a PM2.5 air-quality dataset from North Carolina.

This lab comes after the lessons on **Data Exploration** and **Visual Data Exploration**. The goal is not just to run code, but to decide what to inspect, interpret what you find, and document your reasoning.

## Learning goals

By the end of this lab, you should be able to:

- Load a CSV file into a Pandas dataframe.
- Inspect dataframe structure, column names, data types, and missing values.
- Explore numeric, categorical, and datetime fields.
- Summarize a numeric variable using descriptive statistics.
- Create exploratory visualizations using Pandas plotting methods.
- Compare plot types and choose appropriate visual tools for different questions.
- Write short interpretations of exploratory results.


## Part 1: Set Up the Session

Load the packages and define the folder paths used in this course project.


In [1]:
# Import packages: Path (from pathlib) and Pandas (as pd)
from pathlib import Path
import numpy as np
import pandas as pd

In [2]:
# Set folder paths: project, raw, and processed
project_fldr = Path.cwd().parent
raw_fldr = project_fldr / "data" / "raw" 
processed_fldr = project_fldr / "data" / "processed"

## Part 2: Load the Data
Read `EPAair_PM25_NC2025_raw.csv` into a dataframe named `pm25`.

When reading the file:

- Set `Site ID` to use the `category` data type.
- Parse `Date` as a datetime field.
- Use the date format `%m/%d/%Y`.

>Tip: See section 9 in the **M3-1** notebook.

In [17]:
# Read the CSV file into a dataframe named pm25.
# Set the path to the raw USGS data file
raw_file = raw_fldr / "EPAair_PM25_NC2025_raw.csv"

# Confirm the file exists before trying to import it
print(raw_file)
print("File exists:", raw_file.exists())

# put into data frame
pm25 = pd.read_csv(
    raw_file,
    dtype= {"Site ID": "category"},
    parse_dates=["Date"],
    date_format="%m/%d/%Y"
    
)

pm25.dtypes

c:\Users\mb1012\Desktop\ENVR700\data\raw\EPAair_PM25_NC2025_raw.csv
File exists: True


Date                              datetime64[us]
Source                                       str
Site ID                                 category
POC                                        int64
Daily Mean PM2.5 Concentration           float64
Units                                        str
Daily AQI Value                            int64
Local Site Name                              str
Daily Obs Count                            int64
Percent Complete                         float64
AQS Parameter Code                         int64
AQS Parameter Description                    str
Method Code                              float64
Method Description                           str
CBSA Code                                float64
CBSA Name                                    str
State FIPS Code                            int64
State                                        str
County FIPS Code                           int64
County                                       str
Site Latitude       

In [18]:
# Display the first few rows.
pm25.head()

,Date,Source,Site ID,POC,Daily Mean PM2.5 Concentration,Units,Daily AQI Value,Local Site Name,Daily Obs Count,Percent Complete,...,Method Code,Method Description,CBSA Code,CBSA Name,State FIPS Code,State,County FIPS Code,County,Site Latitude,Site Longitude
0,2025-01-01,AQS,370110002,1,0.7,ug/m3 LC,4,Linville Falls,1,100.0,...,707.0,IMPROVE Module A with Cyclone Inlet-Teflon Fil...,NaN,NaN,37,North Carolina,11,Avery,35.972347,-81.933072
1,2025-01-04,AQS,370110002,1,1.2,ug/m3 LC,7,Linville Falls,1,100.0,...,707.0,IMPROVE Module A with Cyclone Inlet-Teflon Fil...,NaN,NaN,37,North Carolina,11,Avery,35.972347,-81.933072
2,2025-01-07,AQS,370110002,1,1.5,ug/m3 LC,8,Linville Falls,1,100.0,...,707.0,IMPROVE Module A with Cyclone Inlet-Teflon Fil...,NaN,NaN,37,North Carolina,11,Avery,35.972347,-81.933072
3,2025-01-10,AQS,370110002,1,4.2,ug/m3 LC,23,Linville Falls,1,100.0,...,707.0,IMPROVE Module A with Cyclone Inlet-Teflon Fil...,NaN,NaN,37,North Carolina,11,Avery,35.972347,-81.933072
4,2025-01-13,AQS,370110002,1,5.0,ug/m3 LC,28,Linville Falls,1,100.0,...,707.0,IMPROVE Module A with Cyclone Inlet-Teflon Fil...,NaN,NaN,37,North Carolina,11,Avery,35.972347,-81.933072


### Reflection

In one or two sentences, describe what a single row in this dataset appears to represent.

**Your answer:** A single row represents a given air quality measurement at a site in North Carolina, rows seem to be established by the date.


## Part 3: Inspect the Dataframe Structure

Use dataframe attributes and methods to answer the following questions.

- How many rows and columns are in the dataframe?
- What are the column names?
- What data type is assigned to each column?
- Which columns are numeric?
- Which columns are categorical or text-like?
- Which column is datetime?


In [29]:
# How many rows and columns are in the dataframe?
print("There are 10184 rows and 22 columns.")
pm25.shape



There are 10184 rows and 22 columns.


(10184, 22)

In [32]:
# What are the column names?
print("The column names and data types are:")
pm25.columns

The column names and data types are:


Index(['Date', 'Source', 'Site ID', 'POC', 'Daily Mean PM2.5 Concentration',
       'Units', 'Daily AQI Value', 'Local Site Name', 'Daily Obs Count',
       'Percent Complete', 'AQS Parameter Code', 'AQS Parameter Description',
       'Method Code', 'Method Description', 'CBSA Code', 'CBSA Name',
       'State FIPS Code', 'State', 'County FIPS Code', 'County',
       'Site Latitude', 'Site Longitude'],
      dtype='str')

In [33]:
# Display data types and non-null counts for all the columns?
pm25.info()

<class 'pandas.DataFrame'>
RangeIndex: 10184 entries, 0 to 10183
Data columns (total 22 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   Date                            10184 non-null  datetime64[us]
 1   Source                          10184 non-null  str           
 2   Site ID                         10184 non-null  category      
 3   POC                             10184 non-null  int64         
 4   Daily Mean PM2.5 Concentration  10184 non-null  float64       
 5   Units                           10184 non-null  str           
 6   Daily AQI Value                 10184 non-null  int64         
 7   Local Site Name                 10184 non-null  str           
 8   Daily Obs Count                 10184 non-null  int64         
 9   Percent Complete                10184 non-null  float64       
 10  AQS Parameter Code              10184 non-null  int64         
 11  AQS Parameter

### Interpretation

Answer the following based on the output above.

1. How many observations are in the dataset?  
2. How many variables are in the dataset?  
3. Which variable uses a datetime data type?  
4. Why should `Site ID` be stored as a category instead of as a number?  

**Your answers:**


1. 10184 observations.
2. 22 variables
3. Date
4. No math functions should be done on Site ID and this is simply an enumerated list, 1 vs 2 is not simply multiplying.

## Part 4: Check for Missing Values

Missing values are important to identify early because they can affect summaries, plots, and later analysis.


In [35]:
# Count missing values in each column.
missing_counts = pm25.isna().sum()
missing_counts


Date                                 0
Source                               0
Site ID                              0
POC                                  0
Daily Mean PM2.5 Concentration       0
Units                                0
Daily AQI Value                      0
Local Site Name                      0
Daily Obs Count                      0
Percent Complete                     0
AQS Parameter Code                   0
AQS Parameter Description            0
Method Code                         56
Method Description                  56
CBSA Code                         1500
CBSA Name                         1500
State FIPS Code                      0
State                                0
County FIPS Code                     0
County                               0
Site Latitude                        0
Site Longitude                       0
dtype: int64

In [53]:
# Optional: Calculate the percentage of records missing in each column.
# Hint: divide the missing-value counts by the number of rows and multiply by 100.

# Your code here
has_data = pm25.notna().sum()
has_data

missing_percent = missing_counts/(has_data+missing_counts)*100
missing_percent

Date                               0.000000
Source                             0.000000
Site ID                            0.000000
POC                                0.000000
Daily Mean PM2.5 Concentration     0.000000
Units                              0.000000
Daily AQI Value                    0.000000
Local Site Name                    0.000000
Daily Obs Count                    0.000000
Percent Complete                   0.000000
AQS Parameter Code                 0.000000
AQS Parameter Description          0.000000
Method Code                        0.549882
Method Description                 0.549882
CBSA Code                         14.728987
CBSA Name                         14.728987
State FIPS Code                    0.000000
State                              0.000000
County FIPS Code                   0.000000
County                             0.000000
Site Latitude                      0.000000
Site Longitude                     0.000000
dtype: float64

### Interpretation

1. Which columns, if any, contain missing values?  
2. Are the missing values likely to affect the initial exploration? Why or why not?  

**Your answers:**

1. Method Code, Method Description, CBSA Code, & CBSA Name contain missing values. Because these seem to be qualitative data points and not numerical values, I don't think they should impact much of the data?


## Part 5: Explore a Numeric Variable

The main numeric variable in this dataset is daily mean PM2.5 concentration.

Use the column:

```python
"Daily Mean PM2.5 Concentration"
```


In [54]:
# Store the concentration column name in a shorter variable for reuse.
conc_col = "Daily Mean PM2.5 Concentration"



In [ ]:
# Display descriptive statistics for daily mean PM2.5 concentration.
## conc_col is now a shorter variable and is NOT a string so we don't call it with the quotes
pm25[conc_col].describe()


count    10184.000000
mean         7.097840
std          3.700242
min         -1.800000
25%          4.500000
50%          6.500000
75%          9.025000
max         43.700000
Name: Daily Mean PM2.5 Concentration, dtype: float64

In [ ]:
# Compute the mean and median separately.
mean_pm25 = pm25[conc_col].sum()/pm25[conc_col].count()
mean_pm25

median_pm25 = pm25[conc_col].median()
median_pm25


In [64]:
# Print the values
print("Mean PM2.5:", mean_pm25)
print("Median PM2.5:", median_pm25)

Mean PM2.5: 7.097839748625294
Median PM2.5: 6.5


* Use the [quantiles](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.quantile.html) method to display the quantiles of the dataframe. Adjust the command to show the highest and lowest 20th percentile (4.1 and 9.7, respectively).

In [61]:
# Compute selected quantiles.
pm25[conc_col].quantile([0.2, 0.50, 0.75, 0.8])

0.20    4.100
0.50    6.500
0.75    9.025
0.80    9.700
Name: Daily Mean PM2.5 Concentration, dtype: float64

In [66]:
pm25.head()

,Date,Source,Site ID,POC,Daily Mean PM2.5 Concentration,Units,Daily AQI Value,Local Site Name,Daily Obs Count,Percent Complete,...,Method Code,Method Description,CBSA Code,CBSA Name,State FIPS Code,State,County FIPS Code,County,Site Latitude,Site Longitude
0,2025-01-01,AQS,370110002,1,0.7,ug/m3 LC,4,Linville Falls,1,100.0,...,707.0,IMPROVE Module A with Cyclone Inlet-Teflon Fil...,NaN,NaN,37,North Carolina,11,Avery,35.972347,-81.933072
1,2025-01-04,AQS,370110002,1,1.2,ug/m3 LC,7,Linville Falls,1,100.0,...,707.0,IMPROVE Module A with Cyclone Inlet-Teflon Fil...,NaN,NaN,37,North Carolina,11,Avery,35.972347,-81.933072
2,2025-01-07,AQS,370110002,1,1.5,ug/m3 LC,8,Linville Falls,1,100.0,...,707.0,IMPROVE Module A with Cyclone Inlet-Teflon Fil...,NaN,NaN,37,North Carolina,11,Avery,35.972347,-81.933072
3,2025-01-10,AQS,370110002,1,4.2,ug/m3 LC,23,Linville Falls,1,100.0,...,707.0,IMPROVE Module A with Cyclone Inlet-Teflon Fil...,NaN,NaN,37,North Carolina,11,Avery,35.972347,-81.933072
4,2025-01-13,AQS,370110002,1,5.0,ug/m3 LC,28,Linville Falls,1,100.0,...,707.0,IMPROVE Module A with Cyclone Inlet-Teflon Fil...,NaN,NaN,37,North Carolina,11,Avery,35.972347,-81.933072


### Interpretation

1. What is the minimum daily mean PM2.5 concentration?  
2. What is the maximum daily mean PM2.5 concentration?  
3. Is the mean larger or smaller than the median?  
4. What might the relationship between the mean and median suggest about the shape of the distribution?  
5. In what units are PM2.5 concentrations reported?  

**Your answers:**
1. -1.8 (how is it negative?)
2. 43.7
3. The mean is larger than the median.
4. this suggests that the top 50% of the data is very large and has a wider distribution than the bottom 50%.
5. PM2.5 is reported in ug/m3 LC

## Part 6: Explore Categorical Variables

Categorical variables often describe groups, locations, sources, or measurement types.

Start with `Site ID` and `County`.


In [76]:
# How many unique monitoring sites are included? 26
## nunique counts it
pm25["Site ID"].nunique()



26

In [77]:
# List the unique Site ID categories?
pm25["Site ID"].unique()

['370110002', '370210034', '370350004', '370510009', '370570002', ..., '371590021', '371730002', '371730007', '371830014', '371830021']
Length: 26
Categories (26, str): ['370110002', '370210034', '370350004', '370510009', ..., '371730002', '371730007', '371830014', '371830021']

In [79]:
# Count records by county.
pm25["County"].value_counts()


County
Mecklenburg    1226
Forsyth         895
Wake            840
Buncombe        762
New Hanover     709
Catawba         695
Swain           614
Durham          585
Davidson        398
Pitt            362
Guilford        360
Mitchell        359
Edgecombe       358
Cumberland      357
Montgomery      354
Northampton     353
Johnston        352
Rowan           352
Avery            87
Hyde             86
Haywood          80
Name: count, dtype: int64

In [ ]:
# Create a simple bar plot showing the number of records by county.
# Hint: combine value_counts() with Pandas .plot(kind='bar').



### Interpretation

1. How many different monitoring sites are included?  
2. Which source contributes the most observations?  
3. Are observations balanced across sources?  
4. Why might an unbalanced number of observations matter when comparing groups?  

**Your answers:**


## Part 7: Explore the Datetime Field

Datetime fields have useful sub-components such as year, month, day, and day of week.


In [ ]:
# What date range is covered by the dataset? Print the earliest and latest dates in the dataset


In [ ]:
# Count observations by month.


In [ ]:
# Optional: Create a month column for later exploration.


### Interpretation

1. What time period is covered by this dataset?  
2. Does the dataset appear to cover a full calendar year?  
3. Are observations evenly distributed across months?  

**Your answers:**


## Part 8: Visualize the Distribution of PM2.5

Start by visualizing the distribution of daily mean PM2.5 concentrations.


In [ ]:
# Create a histogram of daily mean PM2.5 concentration using Pandas plotting.
# Hint: select the concentration Series and use .plot(kind='hist').



In [ ]:
# Create a box plot of daily mean PM2.5 concentration using Pandas plotting.
# Hint: select the concentration Series and use .plot(kind='box').



### Interpretation

1. Is the distribution symmetric, left-skewed, or right-skewed?  
2. Are there possible outliers?  
3. Which plot made it easier to see the overall shape of the distribution?  
4. Which plot made it easier to identify unusually high values?  

**Your answers:**


## Part 9: Compare PM2.5 by Source

Now compare daily mean PM2.5 concentration across measurement sources.

Create several visualizations, then decide which one is most useful for comparison.

>Below is code to transform the data such that concentration values are separated into different columns based on their source type ("AQS" or "AirNow"). We'll examine this code later, but you will need this format to produce the KDE plot.

In [ ]:
#Transform data so concentration values are in separate columns by Source
pm25_wide = pm25.pivot_table(
    index = ['Site ID','Date'],
    columns = 'Source',
    values = conc_col,
    aggfunc = 'mean'
).reset_index()

#Show the data
pm25_wide.head()

Source,Site ID,Date,AQS,AirNow
0,370110002,2025-01-01,0.7,NaN
1,370110002,2025-01-04,1.2,NaN
2,370110002,2025-01-07,1.5,NaN
3,370110002,2025-01-10,4.2,NaN
4,370110002,2025-01-13,5.0,NaN


In [ ]:
# Create overlaid density (KDE) curves comparing PM2.5 concentration by source. 
# Use Pandas .plot(kind='kde').


In [ ]:
# Create a box plot comparing PM2.5 concentration by source.
# Hint: use Pandas .plot(kind='box', by='Source') or DataFrame.boxplot().



In [ ]:
# Create histograms of PM2.5 concentration for each source.
# Use Pandas plotting to compare the distributions. You may create separate plots or overlay them.



## Part 11: Data Exploration Checklist

Use this checklist when exploring a new dataset.

- [ ] Load the data and confirm it imported correctly.
- [ ] Examine rows, columns, and column names.
- [ ] Inspect data types.
- [ ] Identify missing values.
- [ ] Summarize numeric variables.
- [ ] Examine categorical variables.
- [ ] Explore datetime variables.
- [ ] Visualize distributions.
- [ ] Compare groups when relevant.
- [ ] Document unusual observations.
- [ ] Write short conclusions in plain language.
